From c38a50358528088123531988ca3295bd42d8da7b Mon Sep 17 00:00:00 2001
From: Pawel Socha <pawel.socha@gmail.com>
Date: Fri, 22 Jul 2016 16:35:21 +0200
Subject: [PATCH] fix for caching api version

---
 nova/compute/rpcapi.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/nova/compute/rpcapi.py b/nova/compute/rpcapi.py
index bd451fd..f2f2dec 100644
--- a/nova/compute/rpcapi.py
+++ b/nova/compute/rpcapi.py
@@ -341,6 +341,18 @@ class ComputeAPI(object):
             return LAST_VERSION
         service_version = objects.Service.get_minimum_version(
             context.get_admin_context(), 'nova-compute')
+
+        #https://review.openstack.org/#/c/339072/
+        # NOTE(johngarbutt) when there are no nova-compute services running we
+        # get service_version == 0. In that case we do not want to cache
+        # this result, because we will get a better answer next time.
+        # As a sane default, return the version from the last release.
+        if service_version == 0:
+            LOG.debug("Not caching compute RPC version_cap, because min "
+                  "service_version is 0. Please ensure a nova-compute "
+                  "service has been started. Defaulting to Mitaka RPC.")
+            return self.VERSION_ALIASES["mitaka"]
+
         history = service_obj.SERVICE_VERSION_HISTORY
         try:
             version_cap = history[service_version]['compute_rpc']
-- 
2.7.3

